hydroflow_cli_integration/src/lib.rs

```diff
@@ -393,3 +448,44 @@ where
         self.sink.take().unwrap()
     }
 }
+
+#[pin_project]
+pub struct MergeSource<T, S: Stream<Item = T> + Send + Sync + ?Sized> {
+    marker: PhantomData<T>,
+    #[pin]
+    sources: Vec<Pin<Box<S>>>,
+}
+
+impl<T, S: Stream<Item = T> + Send + Sync + ?Sized> Stream for MergeSource<T, S> {
+    type Item = T;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let sources = &mut self.get_mut().sources;
+        let mut next = None;
+
+        let mut i = 0;
+        while i < sources.len() {
+            match sources[i].as_mut().poll_next(cx) {
+                Poll::Ready(Some(v)) => {
+                    next = Some(v);
+                    break;
+                }
+                Poll::Ready(None) => {
+                    // this happens infrequently, so OK to be O(n)
+                    sources.remove(i);
+                }
+                Poll::Pending => {
+                    i += 1;
+                }
+            }
+        }
```

Comment:
`.retain_mut()` might be a cleaner way to represent this. though maybe slower

Reply:
yeah I wanted to short circuit when you get a ready, which is why I implemented it this way